### DISCUSSION TITLE
SML + HASS + auto discovery: all values do not carry unit

### DISCUSSION THREAD

**Discussion by TheChatty** [arendst/Tasmota#18545] [Q&A] [answered]
### PROBLEM DESCRIPTION
I've got a few Tasmota devices autodiscovered by Home Assistant (from integration [Tasmota](https://www.home-assistant.io/integrations/tasmota) and [fritz](https://www.home-assistant.io/integrations/fritz)).

The power plug measurements are **shown in HA with their correct unit** (e.g. kWh):
<img src=https://user-images.githubusercontent.com/4789510/235349404-f448f433-bbe3-438e-9269-4759013f77cc.png width=300/> <img src=https://user-images.githubusercontent.com/4789510/235349241-43ff38bf-6677-4238-b26c-46931a20a694.png width=200/>

My house meters are read by SML and they **don't carry any unit in HA**:
<img src=https://user-images.githubusercontent.com/4789510/235349129-8b20407a-7ad0-43ff-8826-d78fcbf6c48e.png width=300/> <img src=https://user-images.githubusercontent.com/4789510/235348833-ec34c3d1-dd57-44ae-bc6c-7274cb1ff6c7.png width=200/>

 

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): GoSund SP1 and ESP8266 DevBoard
- [x] Tasmota binary firmware version number used: 12.x
  - [ ] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: esptool.py
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_
- configure SML script including correct units
- have device autodiscovered by Home Assistant
- observe all values carry **no unit**

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
- all values carry **correct unit**

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
See problem description.

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by barbudor** [comment]
Why would it be a Tasmota problem?

**Comment by TheChatty** [comment]
Because non-SML devices are working properly and SML devices do not. Thus I think I made something wrong with Tasmota.

**Comment by arendst** [comment, marked as answer]
No you didn't. The issue is that hatasmota only adds units to predefined names like power, voltage and energy. Your SML names are unknown to hatasmota and therefor you'll need to find a way to tell hass what unit they need.

**Comment by sfromis** [comment]
And Tasmota will only automatically include a unit indication in the payloads in cases where you have a configuration option to select which unit you want to use, like C/F or hPa/mmHg. In other cases, the intention is to always use the same unit for same type of measure across sensors, which often would by default be the most common unit to use, like when humidity is %RH.

**Comment by TheChatty** [comment]
A **workaround** is to define a [template sensor](https://www.home-assistant.io/integrations/template/#unit_of_measurement) in Home Assistant's `configuration.yaml` like this (the name of the auto discovered sensor is in the last line):
```
template:
  - sensor:
      - name: "Stromzähler Verbrauch ganzes Haus"
        unique_id: "StromzHaus"
        unit_of_measurement: 'kWh'
        device_class: "energy"
        state_class: "total_increasing"
        state: >-
            {{ float(states('sensor.energiekonsum_strom_e_in')) | round(3) }}
```

But this causes unnecessary data duplication (if complete sensor domain is included in InfluxDB) and does not provide historical values (from real sensor to template sensor).

@arendst As you said the problem is within the [sensor.py](https://github.com/emontnemery/hatasmota/blob/master/hatasmota/sensor.py#L153) to only add units to predefined names. There's already an [old hatasmota issue](https://github.com/emontnemery/hatasmota/issues/130) asking for improvement.

@sfromis How can I configure to manually include my units in the payload?

@gemu2015 Could the script generate a custom MQTT message including the configured units?

**Comment by sfromis** [comment]
There is no option to add unit specifications to the standard SENSOR payloads. But if you generate your own JSON payloads, you can put whatever there, both from rules, Berry or scripting. Anyway, if hatasmota is to recognize the relation between added unit types (similar to existing units) and specific fields, it would still need modification to recognize the specific fields names. This would not be useful for fields where the unit is always kWh.

 Or a quite different type of payload structure to indicate what field has what unit.

**Comment by TheChatty** [comment]
Do you have an example payload containing unit specifications?

**Comment by sfromis** [comment]
`{"Time":"2023-04-30T16:48:30.417+02:00","BME280":{"Temperature":28.29,"Humidity":29.08,"DewPoint":8.57,"Pressure":1022.68,"SeaPressure":1024.65},"BH1750":{"Illuminance":29},"PressureUnit":"hPa","TempUnit":"C"}`

Nothing in the JSON to establish that "TempUnit" is for "Temperature" and "DewPoint". "PressureUnit" is for "Pressure" and "SeaPressure". Hence just adding further units in the same location is a waste of time, without the relationship being established elsewhere, and still does not make much sense if the unit is invariably "KWh".

As long as the payload format does not separately include a unit specification for each field, the backend (like hatasmota) needs metadata to derive units, no matter if the payload does have units already without these being linked to specific fields.

Also illustrative that the sample has no unit for Illuminance, as this is meant to always be lux. Such implicit units are the norm for all Tasmota sensor fields, except for the few ones where Tasmota has options to choose unit.

**Comment by TheChatty** [comment]
I'm using SML which provides arbitrary units (not invariably "KWh").

The way to go is like this (as I understood it):
- enhance hatasmota to evaluate a new kind of JSON structure including units for measurements (Tasmota would send custom JSON payload in the meantime)
- enhance Tasmota to include units in JSON payload by default

**Comment by Jason2866** [comment]
This will not be done

> enhance Tasmota to include units in JSON payload by default

since it consumes resources in flash and ram and it is a breaking change.

**Reply by TheChatty** [reply]
At least for SML it would be simple to programmatically append `unit` to each `meter metric`. The code and memory impact would be minimal. For my joined three meters it could look like this:
```json
{
  "Time": "2023-04-30T15:48:46",
  "Strom": {
    "P_in%%W": 998.34,
    "L1%%W": 4.29,
    "L2%%W": 645.19,
    "L3%%W": 348.86,
    "E_in%%kWh": 20687.329,
    "E_out%%kWh": 0
  },
"SekHK": {
    "Temp_Outside%%°C": 13,
    "Temp_Flow%%°C": 47,
    "Temp_Return%%°C": 29.7,
    "Temp_Vessel%%°C": 60,
    "Metric_M3H%%l/h": 10,
    "CtrlSig_RK1%%%": 1
  },
  "PriHK": {
    "HeatEnergyE1%%MWh": 70.306,
    "Volume%%m³": 2286.88,
    "PreTemp%%°C": 41.17,
    "AftTemp%%°C": 27.45,
    "Flow%%l/h": 0
  }
}
```
There could also be a different delimiter like double spaces. hatasmota could check for a predefined list of extensions ("%%kWh", "%%MWh", ...) and subtract them from the metrics name and supply the needed unit.

**Comment by nikito7** [comment]
I use old mqtt discovery and UfsRun

https://github.com/nikito7/easyhan/blob/dev/edpbox/eb1-discovery2.txt
